src/test/java/com/squareup/javapoet/JavaFileTest.java

```diff
@@ -154,6 +152,82 @@
         + "}\n");
   }
 
+  @Test public void conflictingChildName() throws Exception {
+    String source = JavaFile.builder("com.squareup.tacos",
+        TypeSpec.classBuilder("A")
+            .addType(TypeSpec.classBuilder("B")
+                .addType(TypeSpec.classBuilder("C")
+                    .addField(ClassName.get("com.squareup.tacos", "A", "Twin", "D"), "d")
+                    .addType(TypeSpec.classBuilder("Twin").build())
+                    .build())
+                .build())
+            .addType(TypeSpec.classBuilder("Twin")
+                .addType(TypeSpec.classBuilder("D")
+                    .build())
+                .build())
+            .build())
+        .build()
+        .toString();
+    assertThat(source).isEqualTo(""
+        + "package com.squareup.tacos;\n"
+        + "\n"
+        + "class A {\n"
+        + "  class B {\n"
+        + "    class C {\n"
+        + "      A.Twin.D d;\n"
+        + "\n"
+        + "      class Twin {\n"
+        + "      }\n"
+        + "    }\n"
+        + "  }\n"
+        + "\n"
+        + "  class Twin {\n"
+        + "    class D {\n"
+        + "    }\n"
+        + "  }\n"
+        + "}\n");
+  }
+
+  @Test public void conflictingNestedChildName() throws Exception {
+    String source = JavaFile.builder("com.squareup.tacos",
+        TypeSpec.classBuilder("A")
+            .addType(TypeSpec.classBuilder("B")
+                .addType(TypeSpec.classBuilder("C")
+                    .addField(ClassName.get("com.squareup.tacos", "A", "Twin", "D"), "d")
+                    .addType(TypeSpec.classBuilder("Nested")
+                        .addType(TypeSpec.classBuilder("Twin").build())
+                        .build())
+                    .build())
+                .build())
+            .addType(TypeSpec.classBuilder("Twin")
+                .addType(TypeSpec.classBuilder("D")
+                    .build())
+                .build())
+            .build())
+        .build()
+        .toString();
+    assertThat(source).isEqualTo(""
+        + "package com.squareup.tacos;\n"
+        + "\n"
+        + "class A {\n"
+        + "  class B {\n"
+        + "    class C {\n"
+        + "      A.Twin.D d;\n"
```

Comment:
I stared for hours at this problem (and first at my generated code, why it did't compile) so i might have made a mistake here. The nested-child case seems unnecessary.